### ISSUE TITLE
New Filters System Specifications

### ISSUE THREAD

**Issue by mbaruh** [python-discord/bot#1530]
This issue documents the planned changes to the filters extension. This needs to be done alongside python-discord/site#479.
Minor details may still change. This issue specifies the full extent of the desired changes, but this can be split into several issues and made into a project.

The changes allow for granular control over the behavior of individual filters, and increase the scope of it. All aspects of the extension, including automatic rules, will be controlled from a single entry-point.

# Filter Settings
Each added filter has settings which dictate what happens if a given message matches it. The settings are:
- Does it ping?
  - Which roles? everyone/here/mods/something else
- Does it delete?
- Does it filter reactions?
- Does it filter usernames and nicknames?
- Does it catch DM's?
- Does it ping for DM's?
- Does it DM the user about the rule? (similarly to what happens for invite links / everyone pings)
  - A secondary configuration for the message sent
- Do the filters apply only in specific channels / categories
- Which roles can bypass the filter?
- Does it apply an infraction? warn / mute / ban
- For domains: be able to choose whether it should be exact, or apply for subdomains as well (this should close #1276)

Each filter additionally has a description, as they already do.

Each list type has default settings, while each individual filter can override any setting.

# Filter Types
The current filters are split between four types: guild_invite, file_format, domain_name, filter_token. Each of those is split into a blacklist and a whitelist.

## Renaming
The names should be simpler and more memorable: invites, formats, domains, tokens.

## Whitelisting and Blacklisting
Not every filter type needs to have both a whitelist and a blacklist. Each filter type has its own method of deciding whether action should be taken, taking the whitelist, the blacklist, or both into account.

A blacklist or a whitelist should be created before anything can be added to it.

## Invites
The filters should save the invite which was used to add a guild to the list, unlike the current state in which only the ID is saved.

## Domains
Domains should support path patterns, e.g `*.pythondiscord.com/*/ric*rolls`.

## New Type: Custom Functions
Each filter type has its own method of checking whether a specific filter should be applied. There should be a list which specifies special filters, each with their own way of deciding whether they apply to a given message. Examples of such filters: the token remover, webhook remover, and any of the antispam rules. When an item can and should be added to this list will be clearer in the `Implementation Details` section.

# UI And Utilities
## Search
There should be an option to present all filters in a specific list which match a given input.
For example, if we have the tokens `a\db` and `123`, and we call `!bl search tokens a4b`, we will be presented with a list which contains the filter `a\db`. The listed filters should show each filter's ID (explained below).

## Aliasing
There should alternative ways to phrase the same command:
`!bl add tokens <token>` -> `!tokens deny <token>`
`!al add invites <invite>` -> `!invites allow <invite>`
`!bl list domains` -> `!domains list bl`

`find` should be an alias of `search`

## Reaction Menus
There should be reaction menus to go from `!bl / !al`, to one of the option: `list, search, add, remove`, which in turn should allow choosing one the possible filter lists (if there is no tokens whitelist it shouldn't be given as an option).

Similarly, following the aliasing section, specifying one of the list types as a top-level command should allow choosing one of `list, search, allow, deny` (if there is no whitelist, `allow` isn't an option).

## Editing
As mentioned in the `Settings` section, each individual filter should have a way to override the default settings, including the pattern or the name of the filter which is used to dictate when it applies to a message. The only non-modifiable field of a filter is its ID (explained below).

## Identification
Each filter should have its own ID, which is unique for a specific list. There should be a way to edit and remove a filter via that ID.

Additionally, it should be possible to bring up an embed with a particular filter's ID, pattern or name, and the values of all of its settings, via either the ID or the pattern / name.

## Alerting
As different filters can have different affects on a message, it should be explored whether it's possible to present in the alert all tokens which triggered unique filters (meaning, we don't have to show all tokens that triggered a specific filter).

## User Ignore List
There should be a command which makes a specific filter temporarily not apply to a specified user. This is the same feature as in #551, except a redis cache might be enough.

# Implementation Details
## OOP-Based Approach
* There should be a `Settings` dataclass that implements a binary operation to join settings together. This will allow us to decide all the actions that should be taken for a specific message. For example, if one filter that is matched says the message should be deleted, and another says the moderators should be pinged, they should be joined together, and then actioned after going over all filters. There should also be a method to apply overrides to a `Settings` object, returning a *new* `Settings` object.

* There should be a base `Filter` class with a settings attribute, that is then subclassed for each filter type to code a predicate to decide whether the filter should be applied (along the lines of `async def matches(message: Message) -> bool`). This will also allow the custom filters to be subclasses of `Filter`.

* Lastly, there should be a base `FilterList` class, that is then subclassed to specify which `Filters` class it uses, how it loads and initializes its filters, and what it does with its blacklist and/or whitelist. Each list loads its filters from the database, transforms each into an object of a `Filter` subclass, and stores it in the list corresponding to the blacklist or whitelist.

## Extensibility and Modularity
I don't claim to foresee every future need or use case. For that reason, during development special attention should be made to extensibility (this also applies to the site issue). Most parts should be modular, with several dispatching methods. A good indication of a hard-to-extend implementation would be if a fragmented diff, scattered across a file, is necessary to say add an additional type of filter, or an additional effect a filter can have.

Furthermore, a module, a file, or a class should have a clear documentation on what parts need to be looked at and added / changed in order to add or remove an additional filter setting or filter type.

Each `FilterList` class should be specified in its own file, within a dedicated module, and all of them should be dynamically loaded in the cog from said module.

Similarly, each custom filter should be in its own file, within a dedicated module. The filter within the custom filters list will then store the name of the class (similarly to how tokens store a regex), which will then be dynamically loaded from said module (as a reminder, each `FilterList` subclass can override how it loads its filters).

# Related Issues
Additionally to #1276 and #551 mentioned above, this specification should allow to gracefully handle #1018, #1421.

#665, #998, #1379, #1522 are additional issues related to the functionality of one or more filters that can be handled as a part of this rewrite, but technically speaking are out of scope for this issue.

**Comment by MarkKoz** [comment]
> Do the filters apply only in specific channels / categories

What about filtering by roles? The obvious use case is to exclude staff members from some filters.

> Reaction Menus

While I'm not the target user for this feature, I'm not sold on the necessity of it.

> it should be explored whether it's possible to present in the alert all tokens which triggered unique filters (meaning, we don't have to show all tokens that triggered a specific filter).

Can you rephrase this? I don't understand what the difference is between the sentence outside the parenthesis and the one inside the parenthesis.

> User Ignore List

What kind of granularity is needed? Specific filter ID(a), filter list type(s), or all filters?

> There should be a base Filter class with a settings attribute, that is then subclassed for each filter type to code a predicate to decide whether the filter should be applied (along the lines of async def matches(message: Message) -> bool). This will also allow the custom filters to be subclasses of Filter.

Is defining subclasses really necessary? I don't see why the predicates would need to be stateful. It could just as easily be done through something like

```py
class Filter(NamedTuple):
    settings: Settings
    predicate: Callable

class FilterRegistry:  # For the sake of example...
    def add_filter(self, settings: Settings, predicate: Callable):
        self.filters.append(Filter(settings, predicate))
```

**Comment by mbaruh** [comment]
> What about filtering by roles? The obvious use case is to exclude staff members from some filters.

You're right; edited.

> While I'm not the target user for this feature, I'm not sold on the necessity of it.

The commands are kinda long and hard to remember, so what ends up happening is that someone will use `!bl`, and then they see the available commands, and then they use say `!bl list`, because we don't remember what the names of the lists are, and only after seeing the list names we'll call the full command `!bl list filter_token` (for example). So that's the same routine of three commands being called just to do the one operation. If there was a way to transition from whatever starting point you remember to the desired output that would be much better. As I'm saying all this I realize that slash commands would do very well here, but I'm not sure we want to go that route yet?

> Can you rephrase this? I don't understand what the difference is between the sentence outside the parenthesis and the one inside the parenthesis.

Both sentences say the same thing, I just tried to make it easier to understand in case someone didn't get the phrasing in the first sentence. The point is that different filters trigger different actions, so unlike now, several might need to be shown to understand the cause for the full set of actions taken. At the same time, we don't want to show the same filter twice.

> Is defining subclasses really necessary? I don't see why the predicates would need to be stateful. It could just as easily be done through something like <...>

The issue is that it's not possible for the list of custom filters, because each such custom filter will have its own predicate, which can be arbitrarily complex. So I'd rather if those classes contained their predicates, and then load them dynamically from a module, instead of the filter list containing all of the predicates and then dispatching them.
While it seems possible to do what you're showing for other list types with non-changing predicates, I'm wondering if it's better to just have the same design for all of them.

**Comment by mbaruh** [comment]
Missed this one

> What kind of granularity is needed? Specific filter ID(a), filter list type(s), or all filters?

Whatever is needed to uniquely identify a filter. By how the site schema looks right now (not sure if Akarys updated it after opening the issue, we continued discussing it since), it seems that each filter will have a unique ID. So a filter ID - user ID's pairing should be enough.

**Comment by mbaruh** [comment]
I want to take a step back for a moment, because I feel like my original description didn't convey my full thoughts properly.

This is how I envision things working:

## Overview
We have the `Filtering` (or however we name it) cog. The cog has an on_message listener, and for each message it asks the question: what should I do with it?

### What should the cog do with the message?
To answer that question, it has an array of filter lists at its service, each represented as an instance of the `FilterList` class or its derivatives. If we port over the current feature-set then we have four lists: `guild_invite`, `file_format`, `domain_name`, `filter_token`. The cog sends each of them the message it received. The filter lists, in turn, have a method of taking a message, and telling the cog what it should do with that message. The cog takes all of the answers, combines them, and actions them. For example, if one list says the message should be deleted, and another says a DM should be sent, both will occur.

### How does each filter list know what to do with a message?
The lists are initialized by the cog on startup. Each list loads a blacklist and / or a whitelist from the database, each being an array of filters. The filters are each initialized with the `Filter` class or its derivatives. Whenever a filter list is given a message, it relays it to each of its filters, and the filters tell the list what should be done with the message. The list then takes all of the answers, combines them, and returns the combined answer to the cog.


### How does each filter know what to do with a message? 
Each filter list has a default set of settings (the kinds of settings are given at the top of this issue). It is default in the sense that, unless a filter says otherwise, if a filter in the list is matched, action should be taken according to the defaults of the list. But, each filter can stray from the defaults and override them. Each filter has a method, let's call it a predicate, to check whether it should trigger on the message. The predicate uses a special token which is unique to the filter (regex for filter_tokens, urls for domain_names, etc.). It is also possible that one or more of the settings will affect the result of the predicate. In conclusion, taking all attributes into account, each filter can check if it should trigger itself, and returns the appropriate actions to take to its list.
___
Then, the answer to this question:
> For this purpose, would a `FilterList` have a single ID or would it be a collection of filters each with a different ID?

Is that the list holds a collection (or two - blacklist, whitelist) of filters, and each filter has its own ID.
___

## Diving into more detail
I want to emphasize that I'm presenting an approach which seems simple to me and that I think makes the code well-structured. I'm sure disadvantages can be found for it like with any approach, but my focus here is making everything modular.

This time let's start from the bottom

### The settings
We have numerous settings, such as:
- Does it ping?
- Does it delete?
- Does it catch DM's?

etc. etc.
There will be a base `SettingEntry` (or however we call it) class. Each type of setting will be represented by a subclass of it. What does this class do?
* It holds the value of the setting. For the question `does it delete?` it will be a boolean. For the question `who does it ping?` it will be a list of ID's.
* `SettingEntry`'s can be combined via an `__or__` method. For `does it delete?`, it will `or` the two values. For `who does it ping?` it will sum the two lists. For `What message to DM?` it can join the messages together in whatever format is chosen.
* The class also defines what action should be taken, using a method, let's call it `execute`. For example, the `DeleteMessage` subclass, given the message, will check if its value is `True`, and if it is will delete the message.
* Another type of entry is one which decides whether any action should be taken at all (for example if the message was sent by a staff member). For this entries will have another method, a predicate, which will return a boolean. 

All of these subclasses will be contained within the same module. All that's left then, is to combine all of these into a single set of settings. For that we have the `Settings` class. An object of this class holds a collection of `SettingEntry` objects, and has its own `execute` function, which simply calls the `execute` of each of its entries. It will additionally have its own `__or__` method which will similarly call `__or__` for each setting entry. And lastly, it will have its own predicate, which will simply check the predicates of its entries.

#### But why like this?
The `Settings` object is initialized by taking a dictionary. The dictionary contains a `setting_entry_name: value` mapping, as it was received from the database. It will look for the `setting_entry_name` in the aforementioned module, and will initialize the object of the matching class using the given value in the mapping.

In other words, if we ever want to add a kind of setting, all we need to do is to add the class to the module. That's literally it. No other file needs to be edited. The `SenttingEntry` classes will take care of their own business, all they need is the message object to take action, and the type of event (message, reaction..) for the predicate. A settings object will dynamically refer to the existing `SettingEntry` classes, and no part of the extension needs to explicitly refer to them. If we ever want to delete the setting, we can simply delete the file the `SettingEntry` subclass is in (assuming nothing else was there), and it will be gone from the extension.

### The filters
We now have functioning settings that can give us information on whether to take action, and to take that action.
Now we need to define the filters that use those settings. The attributes are:
* The ID
* The token
* The text description
* The settings
* An additional attribute that affects how the token is treated depending on the type of filter. For example for a domain filter it will decide whether to match subdomains or not.

The filter then has a predicate that takes the token, the settings, and the additional field, and decides whether it matches. If the settings say the filter shouldn't care about the event (using the settings' predicate described above), then the match attempt is stopped. If it does match, then the settings are returned.

This means that the filter's predicate has a state, it can't simply be given a callable that is disjointed from the other attributes. One could argue you could use the NamedTuple subclass you showed by using a partial function that takes all of those attributes first, and is then stored in the tuple, but then you have a state anyway, it's just function-based instead of object-based.

There is a `Filter` class. We then subclass it to modify what the predicate does with the filter's attributes. The four obvious subclasses are the ones matching the current types of filter: `guild_invite`, `file_format`, `domain_name`, `filter_token`. But there's another angle: This gives us the freedom of making the predicate arbitrarily complex. We could for example port over the anti-spam filter as a subclass of `Filter`.

What do we do with all of those types of filters? we contain them inside a common module, similarly to what we did with the settings entries, the reason for which, other than order, is described below.

### The filter lists
We now have filters that tell us whether they trigger, and if so what action they think should be taken.
Now we need to define the filter lists, which dispatch the event to those filters, and aggregates the results. The attributes are:
* The blacklist (if it exists)
* The whitelist (if it exists)
* The default settings of the list

There is a base `FilterList` class. We then subclass it to create the desired types of lists. The current ones are, again: `guild_invite`, `file_format`, `domain_name`, `filter_token`. Each type of list takes the raw info it's given from the database, and decides how to use it to create its collection(s) of filters. For example, the `filter_token` list type takes the raw list of filter data, represented by dicts from the DB, and sends the contents of the dict to initialize the `Filter` subclass whose predicate does regex matching (with reference to the filter's other attributes).

This gives us a way to create a special type of list: we can create a list whose filters are of various types (`Filter` subclasses). The token in the DB is the name by which the type of filter is decided. Similarly to the setting entries module, the type is dynamically pulled from the aforementioned filter module. For example, a filter against repeated message, or reaction spam, or anything we decide.

In the same way, again, we place the various filter types in a common module.

### The cog
We now have filter lists which tell us what to do with an event. We now need to aggregate the results of those lists in the cog in order to have the actions that need to be taken.

The cog loads the lists from the DB, and by using their name decides which list types fit. It then dynamically initializes the right list type from the filter lists module, and we have a working filtering cog.

If we ever want to add or delete a type of list, all we need to do is to delete the matching file from the filter lists module. Nothing else needs to be edited.

**Comment by MarkKoz** [comment]
> which simply calls the execute of each of its entries

It is not so simple because the order of execution could be important (e.g. it can't DM a user that's banned).

> The class also defines what action should be taken, using a method, let's call it execute.
> 
> Another type of entry is one which decides whether any action should be taken at all (for example if the message was sent by a staff member). For this entries will have another method, a predicate, which will return a boolean.

These could be combined into a single function that takes some sort of `FilterContext` object and modifies it, if needed. This may be more flexible, but I'm not sure if the flexibility is needed. I just wanted to put this idea on the table.

> The dictionary contains a setting_entry_name: value mapping, as it was received from the database.

Why would the settings need to be stored in the database in any capacity? Is this your approach to being able to give individual filters in lists overrides for the default settings?

> The Settings object is initialized by taking a dictionary.

This feels clumsy given the amount of settings there are. I ultimately imagined defining a class that contained the values in class attributes. Then, a metaclass would implicitly wrap those values in `SettingEntry` objects.

> Is that the list holds a collection (or two - blacklist, whitelist) of filters, and each filter has its own ID.

Would there be a case where an entire category of filters needs to be ignored for a user? As you can imagine, manually passing a bunch of IDs to ignore would be tedious. It's also more convenient to permit e.g. posting any link rather than finding the specific filter which would affect the link they were trying to post.

> The token

Is this the value that the filter searches for in the given content? What would be the value of this for more complex filters (like anti-spam and Discord token removers)?

> An additional attribute that affects how the token is treated depending on the type of filter. For example for a domain filter it will decide whether to match subdomains or not.

From an implementation perspective, it makes sense for this to be separate from the other settings. However, it still is basically a setting. Therefore, it could be found to be unintuitive for some settings to be in the `Settings` class and others to be in the `Filter` directly. Is there a better solution? If not, how can the distinction between the two be made clearer?

**Comment by mbaruh** [comment]
> It is not so simple because the order of execution could be important (e.g. it can't DM a user that's banned).

In the schema shown in the site issue, there is a type of setting in a separate table called `filter_action`, which bundles the user-facing actions (it might be better to move the message deletion to a normal setting), so it will handle both DMing the user and issuing the infraction.

> These could be combined into a single function that takes some sort of FilterContext object and modifies it, if needed. This may be more flexible, but I'm not sure if the flexibility is needed. I just wanted to put this idea on the table.

The reason a split might be preferrable is that we will have a lot of filters that will need to be checked, and the current design defers action until we finished going over all of them (to avoid the race conditions we talked about in another issue). That's why we should rule out quickly all of the filters that are irrelevant to the specific event, and only defer what we have to.

> Why would the settings need to be stored in the database in any capacity? Is this your approach to being able to give individual filters in lists overrides for the default settings?

Yes, filters have editable overrides and that needs to be stored somewhere. It should also be noted that the default settings of a list are also editable, so they can't be hardcoded either.

> This feels clumsy given the amount of settings there are. I ultimately imagined defining a class that contained the values in class attributes. Then, a metaclass would implicitly wrap those values in SettingEntry objects.

The idea is that when the cog is loaded, it makes an API call to read all of the filter lists, the filters in each, and the settings of each filter. The cog will take care of sending each filter list only the fetched data that is relevant to that list, and the list in turn will send only the data relevant to the filter. The filter, in turn, has settings it needs to load, so it sends that part of the fetched DB data to the settings init. So that's what I mean by taking a dictionary.

The SettingsEntry objects will indeed be stored as attributes of the settings object. We could do that with a metaclass, but given that the loading is generic it could also just be done with a loop at init.

> Would there be a case where an entire category of filters needs to be ignored for a user? As you can imagine, manually passing a bunch of IDs to ignore would be tedious. It's also more convenient to permit e.g. posting any link rather than finding the specific filter which would affect the link they were trying to post.

I haven't considered ignoring an entire list. I can't imagine it being something we'll want to do, the only time I can think of we might want to ignore an entire category for a user is if they're staff, but that will be included in the settings via their role.
Can you elaborate on the last part? do you mean that you will search for all filters that trigger on a specific message and add the user to their ignore lists? that could be useful, yeah.

> Is this the value that the filter searches for in the given content? What would be the value of this for more complex filters (like anti-spam and Discord token removers)?

Yes. For the filters in the "custom" list (such as anti-spam), it will be the name of the filter, and it will be used to choose the right Filter subclass to initialize. From my previous response:
> This gives us a way to create a special type of list: we can create a list whose filters are of various types (Filter subclasses). The token in the DB is the name by which the type of filter is decided. Similarly to the setting entries module, the type is dynamically pulled from the aforementioned filter module. For example, a filter against repeated message, or reaction spam, or anything we decide.

> From an implementation perspective, it makes sense for this to be separate from the other settings. However, it still is basically a setting. Therefore, it could be found to be unintuitive for some settings to be in the Settings class and others to be in the Filter directly. Is there a better solution? If not, how can the distinction between the two be made clearer?

The distinction is that it helps define what the filter is actually looking for, while the other settings are checking message metadata or dictate what will happen if the filter is triggered.

As you said it can seem as just another setting, which is why it will be completely transparent to the end user. If you pull up the filter's "ID card", you'll see all of the settings, along with the filter's other attributes (the token, the description, etc.). Likewise, editing a setting and editing the description will use the same interface.
